### DISCUSSION TITLE
Converting "NULL" to a Null value

### DISCUSSION THREAD

**Discussion by cisaacson** [rusqlite/rusqlite#1241] [Questions] [answered]
We have are creating a `Vec<String>` for params, and then passing them to `params_from_iter` on the vec and overall it works great. It coerces the type to the proper type in the database based on the type hint (such as `REAL`) when used in a prepared statement for a `REPLACE INTO`. These statements have a fixed number of parameter values. We are currently generating the params in the `Vec` dynamically. We now have a use case to make some of the parameters `Null` values, and obviously a string of `"NULL"` will not work since that is an actual value, not a `Null` value. Given that the `Vec` must contain homogeneous values, is there an easy way to do this? 

I am not sure how we would use this with the `params_from_iter`. The `params_from_iter` takes a `Vec<String>` so it doesn't seem like the right solution.

We use `params_from_iter` because it made it easy to pass the right number of params, but perhaps we need to do it another way (like the `params!` macro with a fixed number of params). There are many different types of statements which we store and execute later, using the dynamic number of params required for that statement. using the `params!` macro it does not seem possible to do this.

We could easily make an `enum` with data, a `String` and `Null` type value in 2 variants, and create a `Vec` of those. But then how would we pass those to each prepared statement, when various statements have various numbers of params?

Any suggestions appreciated.

Great product by the way, really easy to use and very fast.

**Comment by trevyn** [comment, marked as answer]
I think using `Option` just works for this.

**Reply by cisaacson** [reply]
Thanks @trevyn. So that means we could just use a `Vec<Option<String>>` as the argument to `params_from_iter` and it would work? I suppose it would coerce a `None` to a `Null` type and `Some` to the proper value for the type needed for the prepared statement. If so that makes things very easy. I will try it.

**Reply by thomcc** [reply]
Yes. You should also be able to use an iterator that yields those, if you don't want to build the Vec in memory.

**Reply by cisaacson** [reply]
Perfect, thanks for the quick help. I will close the discussion topic.
